### PR TITLE
move `tensorboardX` to extra

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

following: https://github.com/Lightning-AI/lightning/issues/9900#issuecomment-1379954572

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -286,9 +286,9 @@ def __init__(
                 Default: ``1.0``.
 
             logger: Logger (or iterable collection of loggers) for experiment tracking. A ``True`` value uses
-                the default ``TensorBoardLogger``. ``False`` will disable logging. If multiple loggers are
-                provided, local files (checkpoints, profiler traces, etc.) are saved in the ``log_dir`` of
-                the first logger.
+                the default ``TensorBoardLogger`` if it is installed otherwise ``CSVLogger``.
```
Comment (lantiga) [inline src/pytorch_lightning/trainer/trainer.py:289]:
```suggestion
                the default ``TensorBoardLogger`` if `tensorboard` or `tensorboardX` are installed otherwise ``CSVLogger``.
```

File: requirements/pytorch/extra.txt [hunk 2]
```diff
@@ -7,3 +7,4 @@ omegaconf>=2.0.5, <2.3.0
 hydra-core>=1.0.5, <1.3.0
 jsonargparse[signatures]>=4.18.0, <4.19.0
 rich>=10.14.0, !=10.15.0.a, <13.0.0
+tensorboardX>=2.2, <=2.5.1  # min version is set by torch.onnx missing attribute
```
Comment (carmocca) [inline requirements/pytorch/extra.txt:10]: Since it's not a base dependency anymore maybe this should be `tensorboard` again

File: src/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 3]
```diff
@@ -57,9 +62,16 @@ def configure_logger(self, logger: Union[bool, Logger, Iterable[Logger]]) -> Non
...
-            self.trainer.loggers = [
-                TensorBoardLogger(save_dir=self.trainer.default_root_dir, version=SLURMEnvironment.job_id())
-            ]
+            if _TENSORBOARD_AVAILABLE or _TENSORBOARDX_AVAILABLE:
+                logger_ = TensorBoardLogger(save_dir=self.trainer.default_root_dir, version=SLURMEnvironment.job_id())
+            else:
+                warning_cache.warn(
+                    "We are setting default logger but you have not installed preferred `tensorboard`,"
+                    " so we will fall bck to `CSVLogger` instead. If you want to continue with TB,"
+                    " please install this package with `[extra]` options."
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:71]:
```suggestion
                    "`Trainer(logger=True)` is set but `tensorboard` or `tensorboardX` is not available."
                    " You can pip install either, or this package with the `[extra]` option."
                    " We will fall back to `CSVLogger` instead."
```

File: src/pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (carmocca) [inline src/pytorch_lightning/trainer/trainer.py:289]:
```suggestion
                the default ``TensorBoardLogger`` if it is installed, otherwise ``CSVLogger``.
```

File: src/pytorch_lightning/CHANGELOG.md [hunk 4]
```diff
@@ -35,6 +35,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - The Trainer now raises an error if it is given multiple stateful callbacks of the same time with colliding state keys ([#15634](https://github.com/Lightning-AI/lightning/pull/15634))
 - `MLFlowLogger` now logs hyperparameters and metrics in batched API calls ([#15915](https://github.com/Lightning-AI/lightning/pull/15915))
 - Overriding the `on_train_batch_{start,end}` hooks in conjunction with taking a `dataloader_iter` in the `training_step` no longer errors out and instead shows a warning ([#16062](https://github.com/Lightning-AI/lightning/pull/16062))
+- move `tensorboardX` to extra dependencies ([#16349](https://github.com/Lightning-AI/lightning/pull/16349))
```
Comment (carmocca) [inline src/pytorch_lightning/CHANGELOG.md:38]:
```suggestion
- Move `tensorboardX` to extra dependencies. Use the `CSVLogger` by default ([#16349](https://github.com/Lightning-AI/lightning/pull/16349))
```

File: src/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 5]
```diff
@@ -57,9 +62,16 @@ def configure_logger(self, logger: Union[bool, Logger, Iterable[Logger]]) -> Non
...
         elif logger is True:
             # default logger
-            self.trainer.loggers = [
-                TensorBoardLogger(save_dir=self.trainer.default_root_dir, version=SLURMEnvironment.job_id())
-            ]
+            if _TENSORBOARD_AVAILABLE or _TENSORBOARDX_AVAILABLE:
+                logger_ = TensorBoardLogger(save_dir=self.trainer.default_root_dir, version=SLURMEnvironment.job_id())
+            else:
+                warning_cache.warn(
+                    "We are setting default logger but you have not installed preferred `tensorboard`,"
```
Comment (lantiga) [inline src/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:69]:
```suggestion
                    "Starting from release 1.9, `tensorboard` has been removed as a dependency of the `lightning` package, due to potential conflicts with other packages in the ML ecosystem. For this reason, `logger=True` will use `CSVLogger` as the default logger, unless the `tensorboard` package is found. Please `pip install lightning[extra]` to enable TensorBoard support by default".
```

**Review by lantiga — APPROVED** [review]
Tweaked the error message, I hope it provides more context while being easy to understand.

File: requirements/pytorch/extra.txt (hunk 2 above)
Comment (lantiga) [reply requirements/pytorch/extra.txt:10]: yeah why not

File: src/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 5 above)
Comment (lantiga) [reply src/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:69]: @carmocca WDYT about this more verbose explanation?

File: src/pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (lantiga) [reply src/pytorch_lightning/trainer/trainer.py:289]: If we go with `tensorboard` in the extras then we remove the reference to `tensorboardX` here

File: src/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 5 above)
Comment (carmocca) [reply src/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:69]: LGTM

**Review by carmocca — COMMENTED** [review]
I updated the docs

File: requirements/pytorch/extra.txt (hunk 2 above)
Comment (lantiga) [reply requirements/pytorch/extra.txt:10]: `tensorboard` is even worse than `tensorboardX` because it depends on `grpcio`, which required compilation on M1/M2 with Python >=3.10, and still does (tested now)
so there may still be a rationale for going with `tensorboardX` for extras

File: docs/source-pytorch/extensions/logging.rst [hunk 6]
```diff
@@ -79,7 +80,7 @@ Choose from any of the others such as MLflow, Comet, Neptune, WandB, etc.
 To use multiple loggers, simply pass in a ``list`` or ``tuple`` of loggers.
 
 .. testcode::
-    :skipif: not _COMET_AVAILABLE
+    :skipif: not _TENSORBOARD_AVAILABLE and not _TENSORBOARDX_AVAILABLE and not _COMET_AVAILABLE
```
Comment (carmocca) [inline docs/source-pytorch/extensions/logging.rst:83]:
```suggestion
    :skipif: (not _TENSORBOARD_AVAILABLE and not _TENSORBOARDX_AVAILABLE) or not _COMET_AVAILABLE
```

File: docs/source-pytorch/visualize/supported_exp_managers.rst [hunk 7]
```diff
@@ -179,7 +179,7 @@ Use multiple exp managers
 To use multiple experiment managers at the same time, pass a list to the *logger* :class:`~pytorch_lightning.trainer.trainer.Trainer` argument.
 
 .. testcode::
-    :skipif: not _WANDB_AVAILABLE
+    :skipif: not _TENSORBOARD_AVAILABLE and not _TENSORBOARDX_AVAILABLE and not _WANDB_AVAILABLE
```
Comment (carmocca) [inline docs/source-pytorch/visualize/supported_exp_managers.rst:182]:
```suggestion
    :skipif: (not _TENSORBOARD_AVAILABLE and not _TENSORBOARDX_AVAILABLE) or not _WANDB_AVAILABLE
```

File: src/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 8]
```diff
@@ -57,9 +62,17 @@ def configure_logger(self, logger: Union[bool, Logger, Iterable[Logger]]) -> Non
...
         elif logger is True:
             # default logger
-            self.trainer.loggers = [
-                TensorBoardLogger(save_dir=self.trainer.default_root_dir, version=SLURMEnvironment.job_id())
-            ]
+            if _TENSORBOARD_AVAILABLE or _TENSORBOARDX_AVAILABLE:
+                logger_ = TensorBoardLogger(save_dir=self.trainer.default_root_dir, version=SLURMEnvironment.job_id())
+            else:
+                warning_cache.warn(
+                    "Starting from release 1.9, `tensorboard` has been removed as a dependency of the `lightning`"
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:69]:
```suggestion
                    "Starting from v1.9.0, `tensorboardX` has been removed as a dependency of the `pytorch_lightning`"
```

File: src/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 9]
```diff
@@ -57,9 +62,17 @@ def configure_logger(self, logger: Union[bool, Logger, Iterable[Logger]]) -> Non
...
+                    " package, due to potential conflicts with other packages in the ML ecosystem. For this reason,"
+                    " `logger=True` will use `CSVLogger` as the default logger, unless the `tensorboard` package"
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:71]:
```suggestion
                    " `logger=True` will use `CSVLogger` as the default logger, unless the `tensorboard` or `tensorboardX` packages"
```

File: src/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 10]
```diff
@@ -57,9 +62,17 @@ def configure_logger(self, logger: Union[bool, Logger, Iterable[Logger]]) -> Non
...
+                    " is found. Please `pip install lightning[extra]` to enable TensorBoard support by default"
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:72]: My suggestions might need some formatting.
```suggestion
                    " are found. Please `pip install lightning[extra]` or one of them to enable TensorBoard support by default"
```

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 8 above)
Comment (carmocca) [reply src/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:69]: Note that the mirror package will rewrite this as `lightning.pytorch`

**Review by lantiga — APPROVED** [review]
